# HG changeset patch
# User StrikerTheHedgefox <mossj32@gmail.com>
# Date 1743761613 10800
#      Fri Apr 04 07:13:33 2025 -0300
# Node ID 57041369f51ae8b3502c6a321cafff2aa44337aa
# Parent  1d896309cd8588893a1fee07ce38e7081143e125
# EXP-Topic strikers_stuff
Let FakeInventory make use of the "Activation" thing property. Vastly increases its usefulness.

diff -r 1d896309cd85 -r 57041369f51a docs/zandronum-history.txt
--- a/docs/zandronum-history.txt	Fri Apr 04 07:07:02 2025 -0300
+++ b/docs/zandronum-history.txt	Fri Apr 04 07:13:33 2025 -0300
@@ -13,6 +13,7 @@
 
 3.3
 ---
+*+	- FakeInventory actors can now make use of the "Activation" thing property, vastly increasing its utility. Allows the item to target the player collecting the item and vice/versa, or for the item itself to be the activator. [StrikerTheHedgefox]
 *-	- Send inventory that inherits from CustomInventory and Inventory on full update. Fixes issues where weapon checks for inventory fails for late-joining clients, or after a map change. [StrikerTheHedgefox]
 +	- Added A_JumpTo codepointer, which jumps to a state instantly & with prediction. Useful for jumping to an ancestor actor's state without the issues A_Jump with a 256 chance has. [StrikerTheHedgefox]
 +	- Backported A_WeaponOffset from ZDoom Revision 115dbd0, with alterations necessary for Zandronum's current codebase. [StrikerTheHedgefox]
diff -r 1d896309cd85 -r 57041369f51a src/thingdef/olddecorations.cpp
--- a/src/thingdef/olddecorations.cpp	Fri Apr 04 07:07:02 2025 -0300
+++ b/src/thingdef/olddecorations.cpp	Fri Apr 04 07:13:33 2025 -0300
@@ -88,7 +88,17 @@
 			return ( true );
 		}
 
-		INTBOOL success = P_ExecuteSpecial(special, NULL, toucher, false,
+		// [JM] Let's make this thing way more useful through the Activation property.
+		AActor* activatorPtr = toucher;
+
+		if (activationtype & THINGSPEC_ThingTargets)
+			target = toucher;
+		if (activationtype & THINGSPEC_TriggerTargets)
+			toucher->target = this;
+		if ((activationtype & THINGSPEC_ThingActs) && !(activationtype & THINGSPEC_TriggerActs))
+			activatorPtr = this;
+
+		INTBOOL success = P_ExecuteSpecial(special, NULL, activatorPtr, false,
 			args[0], args[1], args[2], args[3], args[4]);
 
 		if (success)
